### PR TITLE
self starting sequence

### PR DESCRIPTION
**Pull request by piotrek6641**
Added new sequence which can start new instances and kill previous one.

### CODE REVIEW

File: js/self-starting-sequence/index.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { ReadableApp } from "@scramjet/types";
+import { defer } from "@scramjet/utility";
+
+const app: ReadableApp<string> = async function(_stream,interval:number) {
+    const hub = this.hub!;
```
Comment (patuwwy) [inline js/self-starting-sequence/index.ts:5]: pls check if this.hub is optional. probably you can remove ! now

File: js/self-starting-sequence/index.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    const instanceId = this.instanceId;
+    const instanceClient = hub.getInstanceClient(instanceId);
+    const seqId = await hub.getInstanceInfo(instanceId).then((res) => { return res.sequence; });
+    const seqClient = hub.getSequenceClient(seqId);
+
+    await defer(interval);
+    this.logger.info(`starting sequence with ${interval} ms interval `);
+    await hub.sendNamedData("info", `starting with id: ${instanceId} +/n`);
```
Comment (patuwwy) [inline js/self-starting-sequence/index.ts:13]: nice but you can write to instance output and set it as topic with given name and content-type

File: js/self-starting-sequence/index.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    await seqClient.start({ appConfig:{} });
```
Comment (patuwwy) [inline js/self-starting-sequence/index.ts:14]: would be nice to pass a param with subsequent instance number :)

File: js/self-starting-sequence/index.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    await instanceClient.kill({ removeImmediately:true });
+    return Promise.resolve("end");
```
Comment (patuwwy) [inline js/self-starting-sequence/index.ts:16]: pls add blank line before return statement

**Review by patuwwy — CHANGES_REQUESTED** [review]
see comments

File: js/self-starting-sequence/index.ts (hunk 1 above)
Comment (piotrek6641) [reply js/self-starting-sequence/index.ts:5]: image

Still having the same issue

**Comment by piotrek6641** [comment]
Thank you @patuwwy, I have made all the changes suggested. Please have a look when you free.

File: js/self-starting-sequence/index.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+const app: ReadableApp<string> = async function(_stream,interval:number,sequenceNumber:number=0) {
+    const hub = this.hub!;
+    const instanceId = this.instanceId;
+    const instanceClient = hub.getInstanceClient(instanceId);
+    const seqId = await hub.getInstanceInfo(instanceId).then((res) => { return res.sequence; });
+    const seqClient = hub.getSequenceClient(seqId);
+
+    await defer(interval);
+    this.logger.info(`starting sequence number ${sequenceNumber} with ${interval} ms interval `);
+    output.push(`starting with id: ${instanceId} +/n`);
```
Comment (patuwwy) [inline js/self-starting-sequence/index.ts:17]:
```suggestion
    output.push(`Starting with id: ${instanceId} \n`);
```

File: js/self-starting-sequence/index.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    await seqClient.start({ appConfig:{interval:interval,sequenceNumber:sequenceNumber+1} });
```
Comment (patuwwy) [inline js/self-starting-sequence/index.ts:18]:
```suggestion
    await seqClient.start({ appConfig: { interval, sequenceNumber: sequenceNumber + 1 }});
```

File: js/self-starting-sequence/index.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    await instanceClient.kill({ removeImmediately:true });
+
+    return output
```
Comment (patuwwy) [inline js/self-starting-sequence/index.ts:21]:
```suggestion
    return output;
```

File: js/self-starting-sequence/index.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+);
+const app: ReadableApp<string> = async function(_stream,interval:number,sequenceNumber:number=0) {
+    const hub = this.hub!;
+    const instanceId = this.instanceId;
+    const instanceClient = hub.getInstanceClient(instanceId);
+    const seqId = await hub.getInstanceInfo(instanceId).then((res) => { return res.sequence; });
+    const seqClient = hub.getSequenceClient(seqId);
+
+    await defer(interval);
+    this.logger.info(`starting sequence number ${sequenceNumber} with ${interval} ms interval `);
```
Comment (patuwwy) [inline js/self-starting-sequence/index.ts:16]:
```suggestion

    this.logger.info(`starting sequence number ${sequenceNumber} with ${interval} ms interval `);
```

**Review by patuwwy — APPROVED** [review]

**Pull request merged**